amazon-kinesis-client/src/test/java/software/amazon/kinesis/lifecycle/ShardConsumerSubscriberTest.java

```diff
@@ -444,4 +448,167 @@ public void cancel() {
         }
     }
 
+    class TestShardConsumerSubscriber extends ShardConsumerSubscriber {
+
+        private int genericWarningLogged = 0;
+        private int readTimeoutWarningLogged = 0;
+        private RecordsPublisher recordsPublisher;
+        private ExecutorService executorService;
+        private int bufferSize;
+        private ShardConsumer shardConsumer;
+
+        TestShardConsumerSubscriber(int readTimeoutsToIgnoreBeforeWarning) {
+            this(mock(RecordsPublisher.class), Executors.newFixedThreadPool(1), 8, mock(ShardConsumer.class),
```

Comment:
AFAIK, @mock only works on test classes, not nest classes within the test. Keeping these within the class aids with OOP and more easily resetting the state of the class rather than worrying about another shared mock that needs to be handled safely and single-threadedly.

Reply:
Unit tests should avoid multithreaded testing. @ Mock normally gets reinitialized between tests. You should not be using the same subscriber between tests, you can make sure to initialize it in the @ Before method with the mocked variable.

Reply:
I'm not using the same subscriber between tests. It's not in the test class scope, which makes it impossible to use the same one between multiple tests. Having test-class-level mocks is what opens up the possibility to accidentally reuse mocks between tests.